### ISSUE TITLE
.match with functions that return Result<., .>

### ISSUE THREAD

**Issue by timvandam** [supermacro/neverthrow#594]
Currently it does not seem possible to handle both `Ok` and `Err` at the same time. Note that this behavior is different from `.andThen(...).orElse(...)` or `.orElse(...).andThen(...)` as the first of the two chained functions may have its return value handled by the second of the two. As far as I am aware there is no neat way of handling this currently. `.match` comes close, but returns a `Promise` rather than `ResultAsync` when applied to `ResultAsync`.

I propose adding the following:
```ts
declare class ResultAsync<T, E> implements PromiseLike<Result<T, E>> {
  fork<A, B>(ok: (t: T) => Result<A, B> | ResultAsync<A, B>, _err: (e: E) => Result<A, B> | ResultAsync<A, B>): ResultAsync<A, B>
}
```

**Comment by timvandam** [comment]
The pattern I have adopted in the meantime is:

```ts
fromSafePromise(x.match(asyncFn1, asyncFn2).then(id))
```

Where `id` is an identity function that serves to flatten the promise returned by the match branch handler

**Comment by paduc** [comment]
Hi @timvandam,

I’m having a hard time understanding what you want to do. Could you provide a more complete example ?

**Comment by timvandam** [comment]
> Hi @timvandam,
> 
> I’m having a hard time understanding what you want to do. Could you provide a more complete example ?

What I’m trying to achieve is behavior analogous to Promise.then with 2 arguments: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then#onrejected. This is not the same as chaining .then and .catch (and similarly .andThen and .orElse can nit be applied to achieve the same behavior in a simple way)

You provide two functions that handle the ok/err cases and return a (Async)Result

**Comment by paduc** [comment]
Thanks @timvandam. Reading [this](https://stackoverflow.com/questions/24662289/when-is-thensuccess-fail-considered-an-antipattern-for-promises) also helped me grasp the topic.

Wouldn't the best solution be to mimick `Promise.then` and accept a second function argument to `Result.andThen` ?

**Comment by timvandam** [comment]
@paduc Yes that would be fine too. I personally would prefer `fork` because its much clearer in my opinion. And neverthrow does not necessarily follow the Promise naming (e.g. by using `orElse` rather than something like `catch`) so I think it would be nicer

**Comment by macksal** [comment]
I like this because it kind of completes the "toolbox".

For mapping to a result:
- andThen (only runs for Ok)
- orElse (only runs for Err)
- fork (provide different paths for Err and Ok)

For mapping to a value/err:
- map (runs for Ok)
- mapErr (runs for Err)
- match (provide different paths for Err and Ok, and also unboxes the value/err)

On the other hand, can you already accomplish the same thing by just returning a ResultAsync from both branches of your **match**?

**Comment by timvandam** [comment]
@macksal achieving the behavior of `fork` by using `andThen` and `orElse` is possible, but not a nice pattern, e.g.:
```ts
resultAsync
  .andThen(async (val) => {
    if (await condition(val)) {
      return ok(val);
    }

        

    return err({ ignore: true, error: new Error('condition not met') });
  })
  .orElse(async (val) => {
    if ('ignore' in val && val.ignore) return err(val.err);

    // normal error handling logic here

  })
```

This is clear from the `Promise` case: in `.then(.).catch(.)`, the `catch` function can capture exceptions thrown in the `then` function. Similarly, in `.catch(.).then(.)`, handling an error in the `catch` function will cause the `then` function to run next. This is clearly different from `.then(a, b)` because `a` and `b` will never both be called. The same difference would apply for `fork`, `andThen`, and `orElse`.

**Comment by macksal** [comment]
@timvandam understood. It works normally for sync but not for async. `fork` seems justifiable to me.

**Comment by paduc** [comment]
In my opinion, it's valuable to have an API with a clear distinction between the Result/ResultAsync "world" (eg andThen, orElse, map, mapErr, ... all methods with a `Result => Result` signature) and the Promise world, for which neverthrow offers entries (eg fromPromise and other `Promise => Result`) and exits (eg unwrap, match and other `Result => Promise`). 

When there is a back and forth between these worlds, it starts to smell like bad code.

I'm all for adding methods like `fork` that helps us keep our code in the Result/ResultAsync side.

**Comment by paduc** [comment]
This is how I normalize:

```ts
const myResultAsync = okAsync().andThen(myUnknownResult);
```

**Comment by dirkluijk** [comment]
It's not that it's not possible with the current API, but it's just a bit cumbersome. I am trying to combine some generic utils (that are not coupled to `neverthrow`) like this:
```ts
myResultAsync
    .orElse(ifInstanceOf(ProblemError, convertProblemError))
    .orElse(ifInstanceOf(HttpStatusError, convertHttpStatusError))
    .orElse(ifInstanceOf(MyCustomError, convertMyCustomError))
    .unwrapOr(response({ message: 'Internal server error' }, { statusCode: 500 }));
```
instead, I need to write:
```ts
myResultAsync
    .orElse(fromThrowable(ifInstanceOf(ProblemError, convertProblemError)))
    .orElse(fromThrowable(ifInstanceOf(HttpStatusError, convertHttpStatusError)))
    .orElse(fromThrowable(ifInstanceOf(MyCustomError, convertMyCustomError)))
    .unwrapOr(response({ message: 'Internal server error' }, { statusCode: 500 }));
```
Which seems a bit of a missed opportunity.

**Comment by andrewhessler** [comment]
> I like this because it kind of completes the "toolbox".
> 
> For mapping to a result:
> 
> * andThen (only runs for Ok)
> * orElse (only runs for Err)
> * fork (provide different paths for Err and Ok)
> 
> For mapping to a value/err:
> 
> * map (runs for Ok)
> * mapErr (runs for Err)
> * match (provide different paths for Err and Ok, and also unboxes the value/err)
> 
> On the other hand, can you already accomplish the same thing by just returning a ResultAsync from both branches of your **match**?

I kind of think `match` muddies the toolbox. Feels like it exists to exist when it's equivalent to `unwrapOr` or `mapErr` if you need the error.

But then I was also thinking, if match is going to exist, there should be a `flatMap` equivalent. `fork` sounds good to me. 🤷
